### PR TITLE
PROJQUAY-880: Add Monitoring component to Quay Operator

### PR DESCRIPTION
**Pull request by syed**
**Issue:** https://issues.redhat.com/browse/PROJQUAY-880

**Changelog:** 

* Added a new `monitoring` component to the Quay Spec
* Added `Service` and `ServiceMonitor` for exporting Quay metrics to the in-cluster Prometheus
* Added  Grafana dashboard with default Graphs for viewing metrics realtime
* Added an Alert which triggers when there are too many crashes of the Quay pod
* Updated `ComponentKind` to use an enum instead of string
* Updated README docs for development 

**Testing:** 

1. Setup a new `QuayRegistry` 

```
apiVersion: quay.redhat.com/v1
kind: QuayRegistry
metadata:
  name: syed-test-quay
  namespace: syed-test-quay
spec:
  components:
    - kind: monitoring
      managed: true
    - kind: clair
      managed: false
```

2. Make sure that you can see `Quay` under the dashboards section on the Openshift Console.

Screen Shot 2021-03-18 at 9 40 38 AM

3. Create an `Org` in your Quay instance and some traffic to generate metrics. The following script should be able to generate enough traffic to display metrics on the dashboard

```
#!/bin/bash

set -x

ORG=$1

export OCP_RELEASE="4.7.2-x86_64"
export LOCAL_REGISTRY='<your-quay-instance>:443'
export LOCAL_REPOSITORY='$ORG/openshift4'
export PRODUCT_REPO='openshift-release-dev'
export LOCAL_SECRET_JSON="$PWD/pull-secrets-fixed.json"
export RELEASE_NAME="ocp-release"

oc adm -a ${LOCAL_SECRET_JSON} release mirror \
     --from=quay.io/${PRODUCT_REPO}/${RELEASE_NAME}:${OCP_RELEASE} \
     --to=${LOCAL_REGISTRY}/${LOCAL_REPOSITORY} \
     --to-release-image=${LOCAL_REGISTRY}/${LOCAL_REPOSITORY}:${OCP_RELEASE}
```

You need to get the `pull-secrets` JSON file from [OpenShift Container Platform downloads](https://access.redhat.com/downloads/content/290/) page and add the credentials to your Quay instance in that file. More information can be found [Here](https://docs.openshift.com/container-platform/4.2/installing/install_config/installing-restricted-networks-preparations.html)

Update the variables to reflect your setup and save the script as `mirror.sh` and run it using the following command:

```
./mirror.sh <your-org-name>
```

Once you start running the above script. You should see some data in the dashboard

Screen Shot 2021-03-13 at 8 27 39 AM

Screen Shot 2021-03-13 at 8 30 31 AM

### CODE REVIEW

File: docs/development.md [hunk 1]
```diff
@@ -6,6 +6,12 @@ The local development experience of the Quay Operator should be easy and straigh
 
 1. Ensure that the `KUBECONFIG` environment variable is set for your Kubernetes cluster (same file that `kubectl` uses).
 
+2. Install the CRDs for the operator
+
+```sh
+$ kubectl apply -f ./deploy/manifests/quay-operator/3.4.0/*.crd.yaml
```
Comment (alecmerdler) [inline docs/development.md]: Change this to the `0.0.1` directory instead.

File: docs/development.md [hunk 2]
```diff
@@ -30,3 +36,11 @@ $ ngrok http 7071
 ```sh
 $ DEV_OPERATOR_ENDPOINT=http://988e36df98ca.ngrok.io go run main.go --namespace <your-namespace>
 ```
+
+3. Access the config editor locally by port-forwarding the service endpoint
```
Comment (alecmerdler) [inline docs/development.md:43]: Nice, I hadn't thought of that when running without `Routes`.

File: controllers/quay/quayregistry_controller.go [hunk 3]
```diff
@@ -84,6 +92,22 @@ func (r *QuayRegistryReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 	updatedQuay := quay.DeepCopy()
 	quayContext := quaycontext.NewQuayRegistryContext()
 
+	isQuayMarkedToBeDeleted := quay.GetDeletionTimestamp() != nil
```
Comment (alecmerdler) [inline controllers/quay/quayregistry_controller.go]: Adding a finalizer seems like a great idea, but can we add it in a separate PR?

File: controllers/quay/quayregistry_controller.go [hunk 4]
```diff
@@ -171,6 +195,13 @@ func (r *QuayRegistryReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return r.reconcileWithCondition(&quay, v1.ConditionTypeRolloutBlocked, metav1.ConditionTrue, v1.ConditionReasonObjectStorageComponentDependencyError, msg)
 	}
 
+	quayContext, updatedQuay, err = r.checkMonitoringAvailable(quayContext, updatedQuay.DeepCopy(), configBundle.Data["config.yaml"])
+	if err != nil && v1.ComponentIsManaged(updatedQuay.Spec.Components, v1.ComponentMonitoring) {
+		msg := fmt.Sprintf("could not check for monitoring support: %s", err)
+
+		return r.reconcileWithCondition(&quay, v1.ConditionTypeRolloutBlocked, metav1.ConditionTrue, v1.ConditionReasonObjectStorageComponentDependencyError, msg)
```
Comment (alecmerdler) [inline controllers/quay/quayregistry_controller.go]: Use `ConditionReasonMonitoringComponentDependencyError` instead.

**Comment by alecmerdler** [comment]
@syed Do you mind cleaning up the PR description and filling out the template?

File: deploy/manifests/quay-operator/0.0.1/quay-operator.clusterserviceversion.yaml [hunk 5]
```diff
@@ -55,6 +55,8 @@ spec:
         - kind: Role
         - kind: Rolebinding
         - kind: HorizontalPodAutoscaler
+        - kind: ServiceMonitor
```
Comment (alecmerdler) [inline deploy/manifests/quay-operator/0.0.1/quay-operator.clusterserviceversion.yaml:59]: Could you add `{"kind": "monitoring", "managed": true}` to the list of components in the `alm-examples` annotation on the CSV?

File: kustomize/components/monitoring/quay-alerts.role.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  - apiGroups:
+      - ""
+    resources:
+      - services
+      - endpoints
+      - pods
+    verbs:
+      - get
+      - list
+      - watch
```
Comment (alecmerdler) [inline kustomize/components/monitoring/quay-alerts.role.yaml]: Add newline.

File: kustomize/components/monitoring/quay-alerts.rolebinding.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: quay-alerts-rolebinding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: quay-alerts-role
+subjects:
+  - kind: ServiceAccount
+    name: prometheus-operator
+    namespace: openshift-monitoring
```
Comment (alecmerdler) [inline kustomize/components/monitoring/quay-alerts.rolebinding.yaml]: Add newline.

File: kustomize/components/monitoring/quay-metrics.servicemonitor.yaml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: quay-metrics-monitor
+spec:
+  endpoints:
+    - port: quay-metrics
+  selector:
+    matchLabels:
+      quay-component: monitoring
```
Comment (alecmerdler) [inline kustomize/components/monitoring/quay-metrics.servicemonitor.yaml]: Add newline.

File: kustomize/components/monitoring/quay-grafana-dashboard.configmap.yaml [hunk 9]
```diff
@@ -0,0 +1,1002 @@
...
+          }
+        ]
+      },
+      "time": {
+        "from": "now-6h",
+        "to": "now"
+      },
+      "timepicker": {},
+      "timezone": "",
+      "title": "Quay",
```
Comment (alecmerdler) [inline kustomize/components/monitoring/quay-grafana-dashboard.configmap.yaml:1020]: This seems problematic if you have multiple `QuayRegistries` in the same namespace. We may need to move the creation of this config into Go in order to add some basic templating...

**Review by alecmerdler — DISMISSED** [review]

**Review by alecmerdler — APPROVED** [review]

**Comment by LiZhang19817** [comment]
What's the default value of monitoring component?
"- kind: monitoring
      managed: true"

**Comment by alecmerdler** [comment]
@LiZhang19817 Yes, the default is `true`.

**Pull request merged**